ratbag/drivers/sinowealth.py

```diff
@@ -0,0 +1,318 @@
+        # already
+        return cls(
+            dpi_disabled_slots=obj.dpi_disabled_slots,
+            report_rate=report_rate,
+            independent_xy_resolution=xy_independent,
+            dpis=dpis,
+        )
+
+
+@attr.s
+class Reply(object):
+    """
+    Object returned from :meth:`Query.run`. This object will have custom
+    attributes depending on the query's reply spec.
+    """
+
+    pass
+
+
+@attr.s
+class Query:
+    report_id: ReportID = attr.ib()
+    reply_report_id: ReportID = attr.ib()
+    cmd: Command = attr.ib()
+    query_spec: List[Spec] = attr.ib()
+    reply_spec: List[Spec] = attr.ib()
+
+    def run(self, rodent: ratbag.driver.Rodent) -> Reply:
+        # Use the long config if available, otherwise the short config
+        if (
+            self.reply_report_id == ReportID.CONFIG
+            and ReportID.CONFIG_LONG in rodent.report_ids["feature"]
+        ):
+            self.reply_report_id = ReportID.CONFIG_LONG
+
+        qspec = [
+            Spec("B", "report_id"),
+            Spec("B", "cmd"),
+        ] + self.query_spec
+
+        query = Parser.from_object(self, qspec, pad_to=self.report_id.size)
+        rodent.hid_set_feature(self.report_id, bytes(query))
+
+        reply_data = rodent.hid_get_feature(self.reply_report_id)
+        # Special handling for config queries:
+        if self.reply_report_id in (ReportID.CONFIG, ReportID.CONFIG_LONG):
+            if len(reply_data) != self.reply_report_id.size:
+                raise ratbag.driver.ProtocolError.from_rodent(
+                    rodent, f"Unexpected reply data size {len(reply_data)}"
+                )
+        # Real error check
+        elif len(reply_data) != len(query):
+            raise ratbag.driver.ProtocolError.from_rodent(
+                rodent, f"Unexpected reply data size {len(reply_data)}"
+            )
+
+        if self.reply_report_id != reply_data[0]:
+            raise ratbag.driver.ProtocolError.from_rodent(
+                rodent,
+                f"Got reply for a different command {reply_data[1]} instead of {self.reply_report_id}",
+            )
+
+        reply = Reply()
+        Parser.to_object(reply_data, self.reply_spec, obj=reply)
+        self.parse_reply(reply)
+        return reply
+
+    def parse_reply(self, reply):
+        """
+        Override this in the subclass if the autoparsing of the Spec fields
+        is insufficient.
+        """
+
+        pass
+
+
+@attr.s
+class QueryFWVersion(Query):
+    @classmethod
+    def create(cls) -> "QueryFWVersion":
+        return cls(
+            report_id=ReportID.CMD,
+            cmd=Command.FW_VERSION,
+            reply_report_id=ReportID.CMD,
+            query_spec=[],
+            reply_spec=[
+                Spec("B", "report_id"),
+                Spec("B", "cmd"),
+                Spec(
+                    "BBBB",
+                    "version",
+                    convert_from_data=lambda s: bytes(s).decode("utf-8"),
```

Comment:
you can now do this as `Spec("4s", "version", convert_from_data=lambda s: s.decode("utf-8"))`, see 4b387f23430395ad9b85ac5f7e2a17311c638e4d. Takes the `int` step out of the `bytes -> ints -> bytes -> string` conversion

Reply:
Nice, thanks.